From 69e23e286cc93518d7c2fe671f9d7d82f36f6084 Mon Sep 17 00:00:00 2001
From: Jose Dapena Paz <jdapena@igalia.com>
Date: Fri, 18 Feb 2022 12:05:43 +0100
Subject: GCC: some GLImplementationParts equal operators cannot be constexpr
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Fix build error in GCC, as the constexpr operator== requires its
invokations to be also constexpr. So the operators with non-trivial
calls should not be constexpr.

GCC build error:
    ../../ui/gl/gl_implementation.h: In member function ‘constexpr bool gl::GLImplementationParts::operator==(gl::ANGLEImplementation) const’:
    ../../ui/gl/gl_implementation.h:74:23: error: temporary of non-literal type ‘gl::GLImplementationParts’ in a constant expression
       74 |     return operator==(GLImplementationParts(angle_impl));
          |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ../../ui/gl/gl_implementation.h:62:18: note: ‘gl::GLImplementationParts’ is not literal because:
       62 | struct GL_EXPORT GLImplementationParts {
          |                  ^~~~~~~~~~~~~~~~~~~~~
    ../../ui/gl/gl_implementation.h:62:18: note:   ‘gl::GLImplementationParts’ is not an aggregate, does not have a trivial default constructor, and has no ‘constexpr’ constructor that is not a copy or move constructor
    ../../ui/gl/gl_implementation.h: In member function ‘constexpr bool gl::GLImplementationParts::operator==(gl::GLImplementation) const’:
    ../../ui/gl/gl_implementation.h:78:23: error: temporary of non-literal type ‘gl::GLImplementationParts’ in a constant expression
       78 |     return operator==(GLImplementationParts(gl_impl));
          |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Bug: 819294
Change-Id: I92f606017d4bc9b83ff06a57d83898685d975a5e
---
 ui/gl/gl_implementation.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/ui/gl/gl_implementation.h b/ui/gl/gl_implementation.h
index 770cb2c8f4f59..42c33f2b97d2d 100644
--- a/ui/gl/gl_implementation.h
+++ b/ui/gl/gl_implementation.h
@@ -70,11 +70,11 @@ struct GL_EXPORT GLImplementationParts {
     return (gl == other.gl && angle == other.angle);
   }
 
-  constexpr bool operator==(const ANGLEImplementation angle_impl) const {
+  bool operator==(const ANGLEImplementation angle_impl) const {
     return operator==(GLImplementationParts(angle_impl));
   }
 
-  constexpr bool operator==(const GLImplementation gl_impl) const {
+  bool operator==(const GLImplementation gl_impl) const {
     return operator==(GLImplementationParts(gl_impl));
   }
 
-- 
2.32.0

